libsolidity/interface/CompilerStack.cpp

```diff
@@ -742,15 +754,15 @@ Scanner const& CompilerStack::scanner(string const& _sourceName) const
 
 SourceUnit const& CompilerStack::ast(string const& _sourceName) const
 {
-	if (m_stackState < ParsingSuccessful)
+	if (m_stackState < ParsingPerformed && !m_parserErrorRecovery)
```

Comment:
Again, why the additional check?

Reply:
This check is the crux of this PR. We want to produce an ast when there are parser errors.

Reply:
Ok, still: There are two potential error cases in this and most of the other functions:
 1. calling this function before the state has progressed to "parsing performed"
 2. calling this functions after "parsing performed", but with errors and without error recovery.

If the state has not yet progressed to "parsing performed", then this should still result in an error even if error recovery is active. What you need is a flag that says that we are in an error state. If error recovery is active, then we can still progress even though we are in an error state. But even if error recovery is active, it is not possible to output any code if the error flag is true.